### PR TITLE
Add text recognition 0014 and 0015 models

### PR DESCRIPTION
**Pull request by morkovka1337**
This PR adds support of two text recognition models. These two models will form a new Pareto-optimal set of text recognition models, replacing text-recognition-0012 & 0013.
New models are evaluated on 5 datasets (same as resnet-fc text recognition).
Text-Recognition-0015 has complicated encoder-decoder architecture with variable number of decoder runs (like in im2latex models). That's why im2latex evaluator was generalized to text-recognition.

- [x] test im2latex models in new evaluator
- [x] test new text recognition models

### CODE REVIEW

File: models/intel/text-recognition-0015/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# formula-recognition-medium-scan-0001 (composite)
```
Comment (Anna-Mironova) [inline models/intel/text-recognition-0015/README.md:1]:
```suggestion
# text-recognition-0015 (composite)
```

File: models/intel/text-recognition-0015/README.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+|--------------------------------------------------------------------|-----------|
+| im2latex_medium_photographed dataset, im2latex-match-images metric | 81.5%     |
+| im2latex_medium_rendered dataset, im2latex-match-images metric     | 95.7%     |
+| Source framework                                                   | PyTorch\* |
+
+Im2latex-match-images metric is calculated by [this algorithm](../../../tools/accuracy_checker/accuracy_checker/metrics/im2latex_images_match.py )
+
+## Encoder model specification
+
+The formula-recognition-medium-scan-0001-encoder model is a ResNeXt-50 like backbone with some initialization layers for decoder
```
Comment (Anna-Mironova) [inline models/intel/text-recognition-0015/README.md:31]:
```suggestion
The text-recognition-0015-encoder model is a ResNeXt-50 like backbone with some initialization layers for decoder
```

File: models/intel/text-recognition-0015/README.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+### Outputs
+
+1.	Name: `hidden`, shape: `1, 512`. Initial context state of the LSTM cell.
+2.	Name: `context`, shape: `1, 512`. Initial hidden state of the LSTM cell.
+3.	Name: `init_0`, shape: `1, 256`. Initial state of the decoder.
+4.	Name: `row_enc_out`, shape: `1, 20, 75, 512`. Features from encoder that are fed to a decoder.
+
+## Decoder model specification
+
+The formula-recognition-medium-scan-0001-decoder model is an LSTM based decoder with attention module.
```
Comment (Anna-Mironova) [inline models/intel/text-recognition-0015/README.md:57]:
```suggestion
The text-recognition-0015-decoder model is an LSTM based decoder with attention module.
```

File: models/intel/text-recognition-0015/accuracy-check.yml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+      datasets:
+        - name: SVT_recognition
+          preprocessing:
+            - type: bgr_to_gray
+              cast_to_float: false
+            - type: resize
+              dst_width: 256
+              dst_height: 64
+          lowercase: true
```
Comment (eaidova) [inline models/intel/text-recognition-0015/accuracy-check.yml:87]: lowercase is part of what?

Comment (morkovka1337) [reply models/intel/text-recognition-0015/accuracy-check.yml:87]: lowercase is part of dataset and goes to evaluator

Comment (morkovka1337) [reply models/intel/text-recognition-0015/accuracy-check.yml:87]: Maybe it would be better to pass to as evaluator parameter (on the same level as `model_type`)

Comment (eaidova) [reply models/intel/text-recognition-0015/accuracy-check.yml:87]: why it is placed between preprocessing and metric?
I think better to mode it before run cofiguration or add as parameter in place, where it is really used (e.g. converter if annotation should be provided in lowercase or adapter if it should be applied to decoded sequence

Comment (morkovka1337) [reply models/intel/text-recognition-0015/accuracy-check.yml]: Moved to evaluator parameters

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py [hunk 5]
```diff
@@ -44,15 +45,17 @@ def from_configs(cls, config):
         preprocessing = PreprocessingExecutor(dataset_config.get('preprocessing', []), dataset.name)
         metrics_executor = MetricsExecutor(dataset_config['metrics'], dataset)
         launcher = create_launcher(config['launchers'][0], delayed_model_loading=True)
+        lowercase = config.get('lowercase', False)
         meta = dataset.metadata
-        model = SequentialModel(
+        model_type = config.get('model_type', 'SequentialFormulaRecognitionModel')
+        model = MODEL_TYPES[model_type](
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py:51]: could you please handle error for case if model_type not in MODEL_TYPES?

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py [hunk 6]
```diff
@@ -264,13 +274,78 @@ def __init__(self, network_info, launcher, models_args, meta, is_blob=None):
...
+
+class SequentialTextRecognitionModel(BaseSequentialModel):
+    def __init__(self, network_info, launcher, models_args, meta, is_blob=None):
+        super().__init__(network_info, launcher, models_args, meta, is_blob=is_blob)
+        self.vocab = network_info['custom_label_map']
+
+    def get_phrase(self, indices):
+        res = ''
+        for idx in indices:
+            if idx != self.eos_index:
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py:305]: do we really need this check?
as I can see idx == eos already filtered in predict method.

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py [hunk 7]
```diff
@@ -264,13 +274,78 @@ def __init__(self, network_info, launcher, models_args, meta, is_blob=None):
...
+
+            dec_res = self.recognizer_decoder.predict(inputs={'features': features,
+                                                              'hidden': dec_state,
+                                                              'decoder_input': tgt,
+                                                              })
+
+            dec_state = dec_res['decoder_hidden']
+            logit = dec_res['decoder_output']
+            logits.append(logit)
+            tgt = np.array([[np.argmax(np.array(logit), axis=1)]])
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py:332]: why do we need double square brackets here?

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py (hunk 6 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py]: The filtering was applied after adding logit to result tensor, changed this.

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py (hunk 7 above)
Comment (morkovka1337) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py]: removed

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py [hunk 8]
```diff
@@ -264,13 +274,72 @@ def __init__(self, network_info, launcher, models_args, meta, is_blob=None):
...
+
+            dec_res = self.recognizer_decoder.predict(inputs={'features': features,
+                                                              'hidden': dec_state,
+                                                              'decoder_input': tgt,
+                                                              })
+
+            dec_state = dec_res['decoder_hidden']
+            logit = dec_res['decoder_output']
+            tgt = np.argmax(logit, axis=1)
+            if np.squeeze(tgt[0]) == self.eos_index:
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/custom_text_recognition_evaluator.py:327]:
```suggestion
            if self.eos_index in tgt[0]:
```

**Comment by eaidova** [comment]
@morkovka1337, do you have some time estimation about when demo support will be added?
Possibly, better to split this PR on 2 parts and merge model support in AC and docs now without waiting demo support.
It allows us to start model validation (and probably quantization) as soon as possible.

**Comment by morkovka1337** [comment]
@eaidova, if nothing unexpected happens, demo should be ready by the end of this week.
Ok, demo will be in another PR, marking this PR as ready.

**Review by eaidova — APPROVED** [review]

**Review by Anna-Mironova — APPROVED** [review]
